### PR TITLE
Twap: Add twapkeeper wiring

### PR DESCRIPTION
**Pull request by ValarDragon**
## What is the purpose of the change

Extract from #2168 , does the app wiring for twap into the Osmosis app

## Brief Changelog

- Add TwapKeeper to osmosis app
- Add Twap to v11 upgrade
- Add GetPoolDenoms() to gamm keeper

## Testing and Verifying

Wiring across an upgrade boundary is covered by existing test e2e

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? needs changelog entry
  - How is the feature or change documented? not applicable, in spec of subsequent logic PR

### CODE REVIEW

File: app/modules.go [hunk 1]
```diff
@@ -121,6 +123,7 @@ func appModules(
 		params.NewAppModule(*app.ParamsKeeper),
 		app.TransferModule,
 		gamm.NewAppModule(appCodec, *app.GAMMKeeper, app.AccountKeeper, app.BankKeeper),
+		twap.NewAppModule(*app.TwapKeeper),
```
Comment (ValarDragon) [inline app/modules.go:126]: Something cool is that with the structure of keeper and module.go being defined in the same file, no other keepers need to be passed in.

Also AppCodec isn't needed in modules that don't serialize interfaces

File: x/gamm/twap/keeper.go [hunk 2]
```diff
@@ -1,8 +1,27 @@
...
+	paramSpace paramtypes.Subspace
+
+	ammkeeper types.AmmInterface
+}
+
+func NewKeeper(storeKey sdk.StoreKey, transientKey *sdk.TransientStoreKey, paramSpace paramtypes.Subspace, ammKeeper types.AmmInterface) *Keeper {
+	return &Keeper{storeKey: storeKey, transientKey: transientKey, paramSpace: paramSpace, ammkeeper: ammKeeper}
+}
+
+// TODO: make this read from a parameter, or hardcode it.
```
Comment (p0mvn) [inline x/gamm/twap/keeper.go:24]: Let's make an issue please

**Review by p0mvn — COMMENTED** [review]
LGTM

**Comment by p0mvn** [comment]
I can help with creating issues / PRs for TODOs and comments that are not going to be immediately addressed. @ValarDragon do you mind resolving comments for TODOs that you're planning on addressing in follow-ups, please? I can take care of the rest

File: x/gamm/twap/keeper.go [hunk 3]
```diff
@@ -1,8 +1,27 @@
...
+func (k *Keeper) PruneEpochIdentifier(ctx sdk.Context) string {
+	return "daily"
```
Comment (p0mvn) [inline x/gamm/twap/keeper.go:26]: Should it be "day" and not "daily" according to the epochs parameter?

File: app/apptesting/gamm.go [hunk 4]
```diff
@@ -96,3 +96,21 @@ func (s *KeeperTestHelper) PrepareBalancerPoolWithPoolAsset(assets []balancer.Po
 	s.NoError(err)
 	return poolId
 }
+
+func (s *KeeperTestHelper) RunBasicSwap(poolId uint64) {
+	denoms, err := s.App.GAMMKeeper.GetPoolDenoms(s.Ctx, poolId)
+	s.Require().NoError(err)
+
+	swapIn := sdk.NewCoins(sdk.NewCoin(denoms[0], sdk.NewInt(1000)))
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply app/apptesting/gamm.go:104]: The point of this one was to not have to think about anything, I think a second helper that does that would be helpful

File: app/keepers/keepers.go [hunk 5]
```diff
@@ -241,6 +244,12 @@ func (appKeepers *AppKeepers) InitNormalKeepers(
 		appKeepers.AccountKeeper, appKeepers.BankKeeper, appKeepers.DistrKeeper)
 	appKeepers.GAMMKeeper = &gammKeeper
 
+	appKeepers.TwapKeeper = twap.NewKeeper(
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply app/keepers/keepers.go:247]: Yeah we can move it to its own module, the way its written this is really just a folder move

File: x/gamm/twap/keeper.go (hunk 3 above)
Comment (ValarDragon) [reply x/gamm/twap/keeper.go:26]: oh woops, yeah it should

**Pull request merged**